cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,379 @@
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Store the field lineage information.
+   *
+   * @param info the field lineage information
+   */
+  public void addFieldLineageInfo(ProgramRunId programRunId, FieldLineageInfo info) {
+    byte[] rowKey = getChecksumRowKey(info.getChecksum());
+    if (table.get(rowKey).isEmpty()) {
+      Put put = new Put(rowKey);
+      put.add(RAW_OPERATION_MARKER, GSON.toJson(info.getOperations()));
+
+      Map<EndPoint, Set<String>> destinationFields = info.getDestinationFields();
+      for (Map.Entry<EndPoint, Set<String>> entry : destinationFields.entrySet()) {
+        put.add(getFieldColumnKey(entry.getKey()), GSON.toJson(entry.getValue()));
+      }
+
+      addSummary(put, INCOMING_DIRECTION_MARKER, info.getIncomingSummary());
+      addSummary(put, OUTGOING_DIRECTION_MARKER, info.getOutgoingSummary());
+      table.put(put);
+    }
+
+    addFieldLineageInfoReferenceRecords(programRunId, info);
+  }
+
+  private void addSummary(Put put, byte[] direction, Map<EndPointField, Set<EndPointField>> summary) {
+    for (Map.Entry<EndPointField, Set<EndPointField>> entry : summary.entrySet()) {
+      put.add(getSummaryColumnKey(direction, entry.getKey()), GSON.toJson(entry.getValue()));
+    }
+  }
+
+  /**
+   * Add records referring to the common operation record having the given checksum.
+   * Operations represent transformations from source endpoints to the destination endpoints.
+   * From source perspective the operations are added as lineage in outgoing direction, while from
+   * destination perspective they are added as lineage in incoming direction.
+   *
+   * @param programRunId program run for which lineage is to be added
+   * @param info the FieldLineageInfo created by program run
+   */
+  private void addFieldLineageInfoReferenceRecords(ProgramRunId programRunId, FieldLineageInfo info) {
+    // For all the destinations, operations represents incoming lineage
+    for (EndPoint destination : info.getDestinations()) {
+      addOperationReferenceRecord(INCOMING_DIRECTION_MARKER, destination, programRunId, info.getChecksum());
+    }
+
+    // For all the sources, operations represents the outgoing lineage
+    for (EndPoint source : info.getSources()) {
+      addOperationReferenceRecord(OUTGOING_DIRECTION_MARKER, source, programRunId, info.getChecksum());
+    }
+  }
+
+  private void addOperationReferenceRecord(byte[] direction, EndPoint endPoint, ProgramRunId programRunId,
+                                           long checksum) {
+    byte[] rowKey = getOperationReferenceRowKey(direction, endPoint, programRunId);
+    Put put = new Put(rowKey);
+    put.add(CHECKSUM_MARKER, checksum);
+    put.add(PROGRAM_MARKER, GSON.toJson(programRunId));
+    table.put(put);
+  }
+
+  public Set<String> getFields(EndPoint endPoint, long start, long end) {
```

Comment:
All public methods should have javadoc to clearly state the method contract. For non-public method, it would be good to have javadoc as well for non-trivial ones.